src/Interpreters/QueryMetricLog.cpp

```diff
@@ -0,0 +1,180 @@
+                parseQuery(codec_parser, "(Delta(4), ZSTD(1))", 0, DBMS_DEFAULT_MAX_PARSER_DEPTH, DBMS_DEFAULT_MAX_PARSER_BACKTRACKS),
+                "Event time."});
+    result.add({"event_time_microseconds",
+                std::make_shared<DataTypeDateTime64>(6),
+                "Event time with microseconds resolution."});
+    result.add({"query_id",
+                std::make_shared<DataTypeString>(),
+                parseQuery(codec_parser, "(ZSTD(1))", 0, DBMS_DEFAULT_MAX_PARSER_DEPTH, DBMS_DEFAULT_MAX_PARSER_BACKTRACKS),
+                "Query ID."});
+
+    for (const auto & metric : memory_metrics)
+    {
+        auto name = fmt::format("CurrentMetric_{}", CurrentMetrics::getName(metric));
+        const auto * comment = CurrentMetrics::getDocumentation(metric);
+        result.add({std::move(name), std::make_shared<DataTypeInt64>(), comment});
+    }
+
+    for (size_t i = 0, end = ProfileEvents::end(); i < end; ++i)
+    {
+        auto name = fmt::format("ProfileEvent_{}", ProfileEvents::getName(ProfileEvents::Event(i)));
+        const auto * comment = ProfileEvents::getDocumentation(ProfileEvents::Event(i));
+        result.add({std::move(name), std::make_shared<DataTypeUInt64>(), comment});
+    }
+
+    return result;
+}
+
+void QueryMetricLogElement::appendToBlock(MutableColumns & columns) const
+{
+    size_t column_idx = 0;
+
+    columns[column_idx++]->insert(getFQDNOrHostName());
+    columns[column_idx++]->insert(DateLUT::instance().toDayNum(event_time).toUnderType());
+    columns[column_idx++]->insert(event_time);
+    columns[column_idx++]->insert(event_time_microseconds);
+    columns[column_idx++]->insert(query_id);
+    columns[column_idx++]->insert(memory);
+    columns[column_idx++]->insert(background_memory);
+
+    for (size_t i = 0, end = ProfileEvents::end(); i < end; ++i)
+        columns[column_idx++]->insert(profile_events[i]);
+}
+
+void QueryMetricLog::shutdown()
+{
+    stopCollect();
+    SystemLog<QueryMetricLogElement>::shutdown();
+}
+
+void QueryMetricLog::stopCollect()
+{
+    std::lock_guard lock(queries_mutex);
+    for (auto & [query_id, status] : queries)
+        status.task->deactivate();
+}
+
+void QueryMetricLog::startQuery(const String & query_id, TimePoint query_start_time, UInt64 interval_milliseconds)
+{
+    QueryMetricLogStatus status;
+    status.interval_milliseconds = interval_milliseconds;
+    status.next_collect_time = query_start_time + std::chrono::milliseconds(interval_milliseconds);
+
+    const auto & profile_events = CurrentThread::getProfileEvents();
+    for (ProfileEvents::Event i = ProfileEvents::Event(0), end = ProfileEvents::end(); i < end; ++i)
+        status.last_profile_events[i] = profile_events[i].load(std::memory_order_relaxed);
+
+    auto context = getContext();
+    const auto & process_list = context->getProcessList();
+    status.task = context->getSchedulePool().createTask("QueryLog", [this, &process_list, query_id] {
```

Comment:
So essentially we are adding a new thread per query that we want to monitor. I don't know how I feel about that. If you have 200 concurrent queries you'll add an extra 200 threads if you have system.query_metric_log enabled :S

Reply:
At least we should be really clear about the impact this will have in terms of extra resource usage per query

Reply:
>So essentially we are adding a new thread per query that we want to monitor.

If I understood correctly the `BackgroundSchedulePool` logic, that's not the case. `QueryMetricLog` is creating a task that is scheduled to be run by one of the threads available in this common pool. It is not spawning a new thread each query. That'd be indeed not such a great idea and would have more implications as you suggested.

Some context: prior to [this commit](https://github.com/ClickHouse/ClickHouse/pull/66532/commits/befb90b7ca1598b6da3da915e7b85367fae6f3bc) the `QueryMetricLog` was spawning its own thread that had a very similar logic to what `BackgroundSchedulePool` has, with similar condition variables, mutexes and all that. Then, I saw `BackgroundSchedulePool` and realized we'd rather use that implementation instead.